README.md
# GPAW tests runs

with custom eigensolver (instead of default davidson)

Runs here are with:

* M06-L Hybrid Functionals
* Hf-N-C triplet I started "testing" few weeks ago



history.txt
   0 . /app/gpaw/software/venv/bin/activate
   1 gpaw
   2 cd /tmp
   3 ls -lt
   4 python auto.py 08.json 
   5 export LD_PRELOAD=/usr/lib/libgfortran.so.3.0.0 
   6 python auto.py 08.json 
   7 export GPAW=/app/gpaw
   8 python auto.py 08.json 
   9 type python
  10 file /app/gpaw/software/venv/bin/python
  11 apk add fie
  12 apk add file
  13 file /app/gpaw/software/venv/bin/python
  14 file /app/gpaw/software/venv/bin/python3 
  15 file /app/httpd/bin/python3
  16 file /app/httpd/bin/python3.7
  17 ldd  /app/httpd/bin/python3.7
  18 unset LD_PRELOAD
  19 ldd  /app/httpd/bin/python3.7
  20 export LD_PRELOAD=/usr/lib/dd
  21 ldd  /app/httpd/bin/python3.7
  22 export LD_PRELOAD=/usr/lib/libgfortran.so.3.0.0 
  23 ldd  /app/httpd/bin/python3.7
  24 ls -lt LD_PRELOAD=/usr/lib/libgfortran.so.3*
  25 ls -lt /usr/lib/libgfortran.so.3*
  26 export LD_PRELOAD=/usr/lib/libgfortran.so.3
  27 ldd  /app/httpd/bin/python3.7
  28 echo $GPAWDIR
  29 ls -lt
  30 vi 3_35912_P4mm_M06-L_1534182016.txt 
  31 ls -lta
  32 vi 08.json 
  33 python auto.y 08.json 
  34 python auto.py 08.json 
  35 gpaw-setup 
  36 ls
  37 rm *restart *LDA
  38 ls -lt
  39 mkdir /app/gpaw/datasets
  40 python auto.py 08.json 
  41 export GPAW_SETUP_PATH=/app/gpaw/datasets
  42 gpaw-setup -f PBE Hf
  43 gpaw-setup -f PBE C
  44 gpaw-setup -f PBE N
  45 python auto.py 08.json 
  46 ls -lt $GPAW_SETUP_PATH
  47 export GPAW_SETUP_PATH=/app/gpaw/datasets
  48 gpaw-setup -f PBE Hf
  49 ls
  50 ls -lt $GPAW_SETUP_PATH
  51 set
  52 ls -lt /app/gpaw/datasets/
  53 cd /app/gpaw/datasets/
  54 gpaw-setup -f PBE Hf
  55 gpaw-setup -f PBE C
  56 gpaw-setup -f PBE N
  57 cd -
  58 python auto.py 08.json 
  59 df
  60 ls -lt
  61 vi 3_35912_P4mm_M06-L_1534182817.txt
  62 ls -lta
  63 ls -lt *json
  64 vi 08.json 
  65 cp 08.json 09.json
  66 vi 09.json 
  67 python auto.py 09.json 
  68 python auto.py 09.json ls -lt
  69 kill %1
  70 fg
  71 ls -lt
  72 vi 3_35912_P4mm_M06-L_1534236870.txt
  73 ls -lt
  74 diff -c 08.json 09.json 
  75 diff  08.json 09.json 
  76 ls -lt | head
  77 rm 3_35912_P4mm_M06-L_1534239687.txt
  78 vi 3_35912_P4mm_M06-L_1534182817.txt
  79 diff  08.json 09.json 
  80 cp 09.json 10.json
  81 vi 10.json 
  82 python auto.py 10.json 
  83 ls -lt
  84 vi 3_35912_P4mm_M06-L_1534239846.txt 
  85 ls -lt
  86 grep Free *txt
  87 vi auto.py 
  88 vim auto.py 
  89 apk add vim
  90 vim auto.py 
  91 ls -lta
  92 file core.34 
  93 ls -lt
  94 vi auto.py 
  95 vim auto.py 
  96 ls -lt
  97 grep -i eigen *txt
  98 vi 3_35912_P4mm_M06-L_1534182817.txt
  99 vi auto.py 
 100 vim auto.py 
 101 python
 102 vi auto.py 
 103 vim auto.py 
 104 python
 105 ls -lt
 106 vi 3_35912_P4mm_M06-L_1534239846.txt
 107 vi auto.py 
 108 vim auto.py 
 109 ls -lt  | head
 110 python auto.py 10.json 
 111 bg
 112 ls -lt
 113 tail -200f 3_35912_P4mm_M06-L_1534467608.txt 
 114 tail -200 -f 3_35912_P4mm_M06-L_1534467608.txt 
 115 moe 3_35912_P4mm_M06-L_1534467608.txt 
 116 more 3_35912_P4mm_M06-L_1534467608.txt 
 117 vim 3_35912_P4mm_M06-L_1534467608.txt 
 118 tail -200 -f 3_35912_P4mm_M06-L_1534467608.txt 
 119 bg
 120 ls -lt
 121 more 3_35912_P4mm_M06-L_1534239846.txt
 122 ls -lt
 123 vi 3_35912_P4mm_M06-L_1534467608.txt
 124 ls -lt
 125 cat 10.json 
 126 w
 127 ls -lt
 128 vi 10.json 
 129 ls -lt *json
 130 cp 10.json 11.json
 131 vi 11.json 
 132 python auto.py 11.json 
 133 bg
 134 ls -lt
 135 tail -200f 3_35912_P4mm_M06-L_1534787359.txt
 136 tail -200 -f 3_35912_P4mm_M06-L_1534787359.txt
 137 fg
 138 jobs
 139 fg
 140 jobs
 141 ls -lt | head
 142 vim 3_35912_P4mm_M06-L_1534787359.txt
 143 ls -lt
 144 cp 11.json 12.json
 145 vi 12.json
 146 python auto.py 12.json  &
 147 tail -200f 3_35912_Pm_M06-L_1534790014.txt
 148 tail -200 -f 3_35912_Pm_M06-L_1534790014.txt
 149 vi 12.json 
 150 vi  3_35912_Pm_M06-L_1534790014.txt
 151 cp 12.json 13.json
 152 vi 13.json 
 153 python auto.py 13.json  &
 154 ls -lta
 155 git status
 156 top
 157 ls  -lt | head
 158 ls
 159 git status
 160 ls
 161 l
 162 ls
 163 du
 164 ls -lt
 165 ls -lt 13.json
 166 vi 13.json 
 167 ls -lt
 168 set
 169 ls -lt
 170 vi 3_35912_Pm_M06-L_1534797209.txt
 171 vi auto.py 
 172 apk add vim
 173 vim auto.py 
 174 apk add git openssh
 175 ls -lt
 176 git init
 177 vi README.md
 178 git init
 179 git add README.md && git commit -m 'initial readme'
 180  git config --global user.email "antouk@dellelce.com"
 181 git config --global user.name "Antonio Dell'Elce"
 182 git commit -m 'initial readme'
 183 ls -lt
 184 it status
 185 git status
 186 echo '*.pyc' >> .gitignore
 187 echo '*.restart' >> .gitignore 
 188 vim .gitignore 
 189 git status
 190 git add auto.py 
 191 git commit -m 'modified auto.py with custom eigensolver'
 192 git status
 193 echo '*.PBE' >> .gitignore 
 194 git statu
 195 git status
 196 git add .gitignore 
 197 git commit -m 'keep it clean'
 198 git status
 199 git add -A .
 200 git status
 201 git commit -m 'initial results'
 202 ls -lt
 203 git status
 204 ls -lta
 205 ls -lt *.json
 206 vi 13.json 
 207 cp 13.json 14.json
 208 vi 14.json
 209 python auto.py 14.json 
 210 bg
 211 tail -200f 3_35912_Pm_M06-L_1536925153.txt
 212 tail -200 -f 3_35912_Pm_M06-L_1536925153.txt
 213 ls -lt
 214 jobs
 215 cp 14.json 15.json
 216 vi 15.json 
 217 python auto.py 15.json &
 218 top
 219 ls -lt
 220 ls -lt | head
 221 tail -200f 3_31687_Pm_M06-L_1536925764.txt
 222 tail -200 -f 3_31687_Pm_M06-L_1536925764.txt
 223 ls -lt
 224 git status
 225 cp 15.json 16.json
 226 vi 16.json 
 227 python auto.py 16.json &
 228 tail -200 -f 3_27462_Pm_M06-L_1536926083.txt
 229 ls -lt
 230 git status
 231 cp 16.json 17.json
 232 vi 17.json 
 233 python auto.py 17.json &
 234 ls -lt
 235 git status
 236 ls -lt
 237 ls -lt | head
 238 tail -200f 3_27462_Pm_M06-L_1536926083.txt
 239 tail -200 -f 3_27462_Pm_M06-L_1536926083.txt
 240 ls -lt | head
 241 tail -200f 3_27462_Pm_M06-L_1536926248.txt
 242 tail -200 -f 3_27462_Pm_M06-L_1536926248.txt
 243 ls -lt
 244 vi 3_35912_P4mm_M06-L_1534182016.txt

auto.py
'''
   AUTO

   Use JSON as a "Execution specification"
'''

from ase.build import bulk
from ase.io import read
from gpaw import GPAW
from gpaw.utilities import h2gpts
from gpaw.mixer import BroydenMixer
from gpaw.mixer import Mixer

#Spacegroup
import spglib

import json
from pprint import pprint

#id generation
import time

#set proc name
from setproctitle import *
import os

#a better range
import itertools

def frange(start, end=None, inc=1.0):
 "An xrange-like generator which yields float values"
 # imitate range/xrange strange assignment of argument meanings
 if end is None:
  end = start + 0.0     # Ensure a float value for 'end'
  start = 0.0

 assert inc                # sanity check

 for i in itertools.count( ):
  next = start + i * inc

  if (inc>0.0 and next>=end) or (inc<0.0 and next<=end):
    break
  yield next

### CLASSES ###
class execution(object):
 
 def __init__(self, executionInput, xyz=None, stateFile='state.json'):
  '''
  Setup a new execution

  Name:  input file
  xyz:   xyz coordinates (for future automated execution)
  stateFile: state file to use for tracking of executions

  TODO: 

  - only ground state in non-periodic cell supported NOW, more to add!
  - inputs from a dict to better handle automated executions
  '''
  if isinstance(executionInput, str):
   self.inp = self.json(executionInput)
   self.inputFileName = executionInput

  if isinstance(executionInput, dict):
   self.inp = executionInput
   self.inputFileName = 'dict' # TODO: generate meaningful name here (how?)

  self.setup()


 def setup(self):
  '''setup'''

  self.functional = self.inp.get('functional') or 'LDA'
  self.runtype = self.inp.get('runtype') or 'calc'

  self.eigensolver = self.inp.get('eigensolver') or 'cg'

  # this actually can't be none!
  _xyz = self.inp.get('xyz')

  if isinstance(_xyz, str):
   self.inputXYZ = _xyz
   self.atoms = read(self.inputXYZ)
  else:
   # support reading frm a dict (TODO: other type of objects - ?)
   from ase.atoms import Atoms
   from ase.atom  import Atom

   self.atoms = \
    Atoms([Atom(a[0],(a[1],a[2],a[3])) for a in _xyz])

  # TODO: support multiple cell types
  self.atoms.cell=(self.inp['cell']['x'],
                   self.inp['cell']['y'],
                   self.inp['cell']['z'])

  # Mixer setup - only "broyden" supported
  self.mixer = self.inp.get('mixer')

  if self.mixer is None:
   self.mixer = Mixer
  else:
   self.mixer = BroydenMixer

  # Custom ID: allow to trace parameter set using a custom id
  self.custom_id = self.inp.get('custom_id')

  # Periodic Boundary Conditions
  self.atoms.pbc=False

  # This should not be done always
  self.atoms.center()

  self.nbands = self.inp.get('nbands')

  self.gpts = None
  self.setup_gpts()

  self.max_iterations = self.inp.get('maxiter') or 333;

  # Name: to be set by __str__ first time it is run
  self.name = None

  # 
  self.logExtension='txt'
  self.dataExtension='gpw'

  # Timing / statistics
  self.lastElapsed = None

 #
 def get_optimizer(self,name='QuasiNewton'):

  from importlib import import_module

  name = name.lower() 
  _optimMap = \
   {
    'mdmin':       'MDMin',
    'hesslbfgs':   'HessLBFGS',
    'linelbfgs':   'LineLBFGS',
    'fire':        'FIRE',
    'lbfgs':       'LBFGS',
    'lbfgsls':     'LBFGSLineSearch',
    'bfgsls':      'BFGSLineSearch',
    'bfgs':        'BFGS',
    'quasinewton': 'QuasiNewton'
   }
  _class = _optimMap[name]
  _m = import_module('ase.optimize')
  return getattr(_m,_class)

 #
 def setup_gpts(self):
  self.gpts=h2gpts(self.inp.get('spacing') or 0.20,
                   self.atoms.get_cell(),
                   idiv=self.inp.get('idiv') or 8)

 def filename(self, force_id=None, extension='txt'):
  '''to correctly support this:
     stateFile support must be completed
     as it isn't just add "_0" to __str__() for now
  '''
  base = self.__str__()

  if self.custom_id is not None:
   base = base + '_' + self.custom_id

  if force_id is None:
   _fn = base + '_0.' + extension
  else:
   _fn = '{}_{}.{}'.format(base,force_id,extension)
  
  return _fn

 #
 @property
 def xyz(self):
  return self.inp['xyz']

 #
 @property
 def spacegroup(self):
  '''international ID of determined spacegroup'''
  sg = spglib.get_symmetry_dataset(spglib.refine_cell(self.atoms))['international']

  return sg.replace('/',':')

 #
 def __str__(self):
  '''get execution name - to be used for log filename

  Components:

  - Number atoms
  - Volume
  - Spacegroup number
  - Functional name

  '''

  if self.name is None:
   self.name = '{}_{}_{}_{}'.format(len(self.atoms),
                             int(self.atoms.get_volume()*100),
                             self.spacegroup,
                             self.functional)

  _gpaw = os.environ['GPAW']
  _pwd = os.environ['PWD']

  self.procname='{} ({})'.format(self.name, _pwd.replace(_gpaw+'/',''))
  setproctitle(self.procname)

  return self.name

 ##
 def print(self, *args, **kwargs):
  '''print wrapper - initial step before logger'''
  r = print(*args, *kwargs)
  return r

 #
 def pretty(self):
  '''pretty print original parameters'''
  _p=pformat(self.inp)
  self.print(_p)

 #
 def json(self,filename):
  '''load parameters from json'''

  with open(filename,'rb') as f:
   s = f.read()
   j = json.loads(s)

  return j

 #
 def run(self, force_id=None, runtype=None):
  '''
    perform execution - only calc or optim supported now
  '''

  _lastStart = time.time()

  if runtype is None and self.runtype is 'calc':
   _f = self.calc
  else:
   _f = self.optim

  # calculate
  try:
   result = _f(force_id)
  finally:
   _lastEnd = time.time()
   self.lastElapsed = _lastEnd - _lastStart

  return result

 #
 def calc(self, force_id=None, functional=None):
  '''prepare and launch execution
     force_id: use this id for filename generation
  '''

  # Allow to override the funtional
  _fnl = functional or self.functional

  # generate filenames
  _fn = self.filename(force_id)
  _dn = self.filename(force_id, extension=self.dataExtension)

  self.print('Calculation with {} log and Input: {}.'.format(_fn, self.inputFileName))

  if self.nbands is None:
   _calc = GPAW(
                xc=_fnl,
                txt=self.filename(force_id),
                mixer=self.mixer(),
                eigensolver=self.eigensolver,
                maxiter=self.max_iterations,
                gpts=self.gpts
               )
  else:
   _calc = GPAW(
                xc=_fnl,
                txt=self.filename(force_id),
                nbands=self.nbands,
                mixer=self.mixer(),
                eigensolver=self.eigensolver,
                maxiter=self.max_iterations,
                gpts=self.gpts
               )


  self.atoms.set_calculator(_calc)
  self.atoms.get_potential_energy()

  # write out wavefunctions
  _calc.write(_dn, 'all')
  return _calc

 #####
 def optim(self, force_id=None,functional=None):
  '''prepare and launch execution
     force_id: use this id for filename generation
  '''

  _optim_name = self.inp.get('optimizer') or 'QuasiNewton'
  optimizer = self.get_optimizer(_optim_name)

  # Allow to override the funtional
  _fnl = functional or self.functional

  # generate filenames
  _fn = self.filename(force_id)
  _dn = self.filename(force_id, extension=self.dataExtension)

  self.print('Optimization with {} log and Input: {}.'.format(_fn, self.inputFileName))

  if self.nbands is None:
   _calc = GPAW(
                xc=_fnl,
                txt=self.filename(force_id),
                mixer=self.mixer(),
                eigensolver=self.eigensolver,
                maxiter=self.max_iterations,
                gpts=self.gpts
               )
  else:
   _calc = GPAW(
                xc=_fnl,
                txt=self.filename(force_id),
                nbands=self.nbands,
                eigensolver=self.eigensolver,
                mixer=self.mixer(),
                maxiter=self.max_iterations,
                gpts=self.gpts
               )

  self.atoms.set_calculator(_calc)

  self.opt = optimizer(self.atoms, trajectory=self.__str__()+'_emt.traj')
  self.opt.run(fmax=0.05)

  # write out wavefunctions
  _calc.write(_dn, 'all')
  return _calc

###
def main(argv):
 '''basic main function'''

 try:
  inputFile = argv[1]
 except:
  inputFile = 'input.json'

 inp = execution(inputFile)

 print(inp)
 try:
  inp.run(str(int(time.time())))
 except Exception as e:
  print('*EXCEPTION: '+str(e))
 finally:
  if inp.lastElapsed is not None:
   print('  ===> Elpased ' + str(inp.lastElapsed))

## call main

if __name__ == '__main__':
 import sys
 main(sys.argv)

## EOF ##


auto-dzp.py
'''
   AUTO

   Use JSON as a "Execution specification"
'''

from ase.build import bulk
from ase.io import read
from gpaw import GPAW
from gpaw.utilities import h2gpts
from gpaw.mixer import BroydenMixer
from gpaw.mixer import Mixer

#Spacegroup
import spglib

import json
from pprint import pprint

#id generation
import time

#set proc name
from setproctitle import *
import os

#a better range
import itertools

def frange(start, end=None, inc=1.0):
 "An xrange-like generator which yields float values"
 # imitate range/xrange strange assignment of argument meanings
 if end is None:
  end = start + 0.0     # Ensure a float value for 'end'
  start = 0.0

 assert inc                # sanity check

 for i in itertools.count( ):
  next = start + i * inc

  if (inc>0.0 and next>=end) or (inc<0.0 and next<=end):
    break
  yield next

### CLASSES ###
class execution(object):
 
 def __init__(self, executionInput, xyz=None, stateFile='state.json'):
  '''
  Setup a new execution

  Name:  input file
  xyz:   xyz coordinates (for future automated execution)
  stateFile: state file to use for tracking of executions

  TODO: 

  - only ground state in non-periodic cell supported NOW, more to add!
  - inputs from a dict to better handle automated executions
  '''
  if isinstance(executionInput, str):
   self.inp = self.json(executionInput)
   self.inputFileName = executionInput

  if isinstance(executionInput, dict):
   self.inp = executionInput
   self.inputFileName = 'dict' # TODO: generate meaningful name here (how?)

  self.setup()


 def setup(self):
  '''setup'''

  self.functional = self.inp.get('functional') or 'LDA'
  self.runtype = self.inp.get('runtype') or 'calc'

  self.eigensolver = self.inp.get('eigensolver') or 'cg'

  # this actually can't be none!
  _xyz = self.inp.get('xyz')

  if isinstance(_xyz, str):
   self.inputXYZ = _xyz
   self.atoms = read(self.inputXYZ)
  else:
   # support reading frm a dict (TODO: other type of objects - ?)
   from ase.atoms import Atoms
   from ase.atom  import Atom

   self.atoms = \
    Atoms([Atom(a[0],(a[1],a[2],a[3])) for a in _xyz])

  # TODO: support multiple cell types
  self.atoms.cell=(self.inp['cell']['x'],
                   self.inp['cell']['y'],
                   self.inp['cell']['z'])

  # Mixer setup - only "broyden" supported
  self.mixer = self.inp.get('mixer')

  if self.mixer is None:
   self.mixer = Mixer
  else:
   self.mixer = BroydenMixer

  # Custom ID: allow to trace parameter set using a custom id
  self.custom_id = self.inp.get('custom_id')

  # Periodic Boundary Conditions
  self.atoms.pbc=False

  # This should not be done always
  self.atoms.center()

  self.nbands = self.inp.get('nbands')

  self.gpts = None
  self.setup_gpts()

  self.max_iterations = self.inp.get('maxiter') or 333;

  # Name: to be set by __str__ first time it is run
  self.name = None

  # 
  self.logExtension='txt'
  self.dataExtension='gpw'

  # Timing / statistics
  self.lastElapsed = None

 #
 def get_optimizer(self,name='QuasiNewton'):

  from importlib import import_module

  name = name.lower() 
  _optimMap = \
   {
    'mdmin':       'MDMin',
    'hesslbfgs':   'HessLBFGS',
    'linelbfgs':   'LineLBFGS',
    'fire':        'FIRE',
    'lbfgs':       'LBFGS',
    'lbfgsls':     'LBFGSLineSearch',
    'bfgsls':      'BFGSLineSearch',
    'bfgs':        'BFGS',
    'quasinewton': 'QuasiNewton'
   }
  _class = _optimMap[name]
  _m = import_module('ase.optimize')
  return getattr(_m,_class)

 #
 def setup_gpts(self):
  self.gpts=h2gpts(self.inp.get('spacing') or 0.20,
                   self.atoms.get_cell(),
                   idiv=self.inp.get('idiv') or 8)

 def filename(self, force_id=None, extension='txt'):
  '''to correctly support this:
     stateFile support must be completed
     as it isn't just add "_0" to __str__() for now
  '''
  base = self.__str__()

  if self.custom_id is not None:
   base = base + '_' + self.custom_id

  if force_id is None:
   _fn = base + '_0.' + extension
  else:
   _fn = '{}_{}.{}'.format(base,force_id,extension)
  
  return _fn

 #
 @property
 def xyz(self):
  return self.inp['xyz']

 #
 @property
 def spacegroup(self):
  '''international ID of determined spacegroup'''
  sg = spglib.get_symmetry_dataset(spglib.refine_cell(self.atoms))['international']

  return sg.replace('/',':')

 #
 def __str__(self):
  '''get execution name - to be used for log filename

  Components:

  - Number atoms
  - Volume
  - Spacegroup number
  - Functional name

  '''

  if self.name is None:
   self.name = '{}_{}_{}_{}'.format(len(self.atoms),
                             int(self.atoms.get_volume()*100),
                             self.spacegroup,
                             self.functional)

  _gpaw = os.environ['GPAW']
  _pwd = os.environ['PWD']

  self.procname='{} ({})'.format(self.name, _pwd.replace(_gpaw+'/',''))
  setproctitle(self.procname)

  return self.name

 ##
 def print(self, *args, **kwargs):
  '''print wrapper - initial step before logger'''
  r = print(*args, *kwargs)
  return r

 #
 def pretty(self):
  '''pretty print original parameters'''
  _p=pformat(self.inp)
  self.print(_p)

 #
 def json(self,filename):
  '''load parameters from json'''

  with open(filename,'rb') as f:
   s = f.read()
   j = json.loads(s)

  return j

 #
 def run(self, force_id=None, runtype=None):
  '''
    perform execution - only calc or optim supported now
  '''

  _lastStart = time.time()

  if runtype is None and self.runtype is 'calc':
   _f = self.calc
  else:
   _f = self.optim

  # calculate
  try:
   result = _f(force_id)
  finally:
   _lastEnd = time.time()
   self.lastElapsed = _lastEnd - _lastStart

  return result

 #
 def calc(self, force_id=None, functional=None):
  '''prepare and launch execution
     force_id: use this id for filename generation
  '''

  # Allow to override the funtional
  _fnl = functional or self.functional

  # generate filenames
  _fn = self.filename(force_id)
  _dn = self.filename(force_id, extension=self.dataExtension)

  self.print('Calculation with {} log and Input: {}.'.format(_fn, self.inputFileName))

  if self.nbands is None:
   _calc = GPAW(
                mode='fd',
                xc=_fnl,
                basis='dzp', #used only in the initial LCAO step
                txt=self.filename(force_id),
                mixer=self.mixer(),
                eigensolver=self.eigensolver,
                maxiter=self.max_iterations,
                gpts=self.gpts
               )
  else:
   _calc = GPAW(
                mode='fd',
                xc=_fnl,
                txt=self.filename(force_id),
                nbands=self.nbands,
                basis='dzp', #used only in the initial LCAO step
                mixer=self.mixer(),
                eigensolver=self.eigensolver,
                maxiter=self.max_iterations,
                gpts=self.gpts
               )


  self.atoms.set_calculator(_calc)
  self.atoms.get_potential_energy()

  # write out wavefunctions
  _calc.write(_dn, 'all')
  return _calc

 #####
 def optim(self, force_id=None,functional=None):
  '''prepare and launch execution
     force_id: use this id for filename generation
  '''

  _optim_name = self.inp.get('optimizer') or 'QuasiNewton'
  optimizer = self.get_optimizer(_optim_name)

  # Allow to override the funtional
  _fnl = functional or self.functional

  # generate filenames
  _fn = self.filename(force_id)
  _dn = self.filename(force_id, extension=self.dataExtension)

  self.print('Optimization with {} log and Input: {}.'.format(_fn, self.inputFileName))

  if self.nbands is None:
   _calc = GPAW(
                xc=_fnl,
                txt=self.filename(force_id),
                mixer=self.mixer(),
                eigensolver=self.eigensolver,
                maxiter=self.max_iterations,
                gpts=self.gpts
               )
  else:
   _calc = GPAW(
                xc=_fnl,
                txt=self.filename(force_id),
                nbands=self.nbands,
                eigensolver=self.eigensolver,
                mixer=self.mixer(),
                maxiter=self.max_iterations,
                gpts=self.gpts
               )

  self.atoms.set_calculator(_calc)

  self.opt = optimizer(self.atoms, trajectory=self.__str__()+'_emt.traj')
  self.opt.run(fmax=0.05)

  # write out wavefunctions
  _calc.write(_dn, 'all')
  return _calc

###
def main(argv):
 '''basic main function'''

 try:
  inputFile = argv[1]
 except:
  inputFile = 'input.json'

 inp = execution(inputFile)

 print(inp)
 try:
  inp.run(str(int(time.time())))
 except Exception as e:
  print('*EXCEPTION: '+str(e))
 finally:
  if inp.lastElapsed is not None:
   print('  ===> Elpased ' + str(inp.lastElapsed))

## call main

if __name__ == '__main__':
 import sys
 main(sys.argv)

## EOF ##
